articles/build-a-ticketing-app-with-adonisjs-and-vuejs/index.md

```diff
@@ -0,0 +1,902 @@
+        if (event === null) {
+          return response.status(404).json({ message: 'Event is not valid' })
+        }
+        const user = await auth.authenticate()
+        // Check if price matches
+        if (event.ticket_price != request.input('amount')) {
+          const message =
+            'Ticket with id: ' +
+            event.id +
+            ' with amount: ' +
+            event.ticket_price +
+            ' does not equal to User amount: ' +
+            request.input('amount')
+          return response.status(422).json({ message })
+        }
+        const ticket = new Ticket()
+        ticket.userId = user.id
+        ticket.eventId = event.id
+        ticket.amount = request.input('amount')
+        ticket.code = Keygen.hex(5)
+        if (ticket.save()) {
+          // Send User Email, Send Code
+          await User.find(user.id)
+          // user.notifyNow(new TicketNotification(ticket, event));
+          return response
+            .status(200)
+            .json({ message: 'Payment for event with id: ' + event.id + ' was successful' })
+        }
+        return response.status(500).json({ message: 'Internal Server Error, Please try again' })
+      }
+      private async findEvent(id: number): Promise<Event | null> {
+        try {
+          const event = await Event.find(id)
+          if (event) {
+            await event.preload('user')
+            await event.preload('tickets')
+            return event
+          }
+        } catch (error) {
+          console.log(error)
+        }
+        return null
+      }
+    }
+```
+
+The controller includes all the possible CRUD functionalities such as retrieving, storing, updating and deleting the event. It also includes `buy` and `join` method which is use to order and join the ticket for an event.
+
+#### Creating endpoint routes
+
+In this step, we will create all the routes for this project once by adding the following code to our `start/routes.ts` file.
+
+```js
+Route.group(() => {
+  Route.group(() => {
+    Route.post("register", "AuthController.register");
+    Route.post("login", "AuthController.login");
+  }).prefix("auth");
+  Route.group(() => {
+    Route.resource("events", "EventsController").apiOnly();
+    Route.resource("tickets", "TicketsController").apiOnly();
+    Route.post("events/buy/:id", "EventsController.buy");
+    Route.post("events/join/:id", "EventsController.join");
+  }).middleware("auth:api");
+}).prefix("api/v1");
+```
+
+We created three different route groups with a different prefix assigned to each while the last group has the `auth` middleware applied to make sure the request is authenticated before it can access those endpoints.
+
+You can test out your API immediately by using either [Postman](https://www.postman.com/) client or [Hoppscotch](https://hoppscotch.io/) on your browser before developing the frontend.
+
+![Test the API](/engineering-education/build-a-ticketing-app-with-adonisjs-and-vuejs/buildpreview.png)
```

Comment:
Add an image footer describing about the image, something like this:

*Testing the API*